lib/hutch/broker.rb

```diff
@@ -215,41 +216,8 @@ def nack(delivery_tag)
       channel.nack(delivery_tag, false, false)
     end
 
-    def log_publication(serializer, payload, routing_key)
-      logger.info {
-        spec =
-          if serializer.binary?
-            "#{payload.bytesize} bytes message"
-          else
-            "message '#{payload}'"
-          end
-        "publishing #{spec} to #{routing_key}"
-      }
-    end
-
-    def publish(routing_key, message, properties = {}, options = {})
-      ensure_connection!(routing_key, message)
-
-      serializer = options[:serializer] || @config[:serializer]
-
-      non_overridable_properties = {
-        routing_key:  routing_key,
-        timestamp:    connection.current_timestamp,
-        content_type: serializer.content_type,
-      }
-      properties[:message_id]   ||= generate_id
-
-      payload = serializer.encode(message)
-
-      log_publication(serializer, payload, routing_key)
-
-      response = exchange.publish(payload, {persistent: true}.
-        merge(properties).
-        merge(global_properties).
-        merge(non_overridable_properties))
-
-      channel.wait_for_confirms if @config[:force_publisher_confirms]
-      response
+    def publish(*args)
+      publisher.new(connection, channel, exchange, @config).publish(*args)
```

Comment:
Well, you'll want to at least wait to initialise the publisher until connection, channel, and exchange are all set.

My thoughts were toward initialising the publisher object right before publishing, using memoization to ensure it's only done once. It sounds like you'd prefer to avoid memoization, and to initialise the publisher immediately after setting those properties.  I'll make those changes now (and override my memoization changes)